### PR TITLE
[lldb] Add frame recognizer for __builtin_verbose_trap

### PR DESCRIPTION
**Pull request by Michael137**
This patch adds a frame recognizer for Clang's
`__builtin_verbose_trap`, which behaves like a
`__builtin_trap`, but emits a failure-reason string into debug-info in order for debuggers to display
it to a user.

The frame recognizer triggers when we encounter
a frame with a function name that begins with
`__llvm_verbose_trap`, which is the magic prefix
Clang emits into debug-info for verbose traps.
Once such frame is encountered we display the
frame function name as the `Stop Reason` and display that frame to the user.

Example output:
```
(lldb) run
warning: a.out was compiled with optimization - stepping may behave oddly; variables may not be available.
Process 35942 launched: 'a.out' (arm64)
Process 35942 stopped
* thread #1, queue = 'com.apple.main-thread', stop reason = __llvm_verbose_trap: Function is not implemented
    frame #1: 0x0000000100003fa4 a.out`main [inlined] Dummy::func(this=<unavailable>) at verbose_trap.cpp:3:5 [opt]
   1    struct Dummy {
   2      void func() {
-> 3        __builtin_verbose_trap("Function is not implemented");
   4      }
   5    };
   6
   7    int main() {
(lldb) bt
* thread #1, queue = 'com.apple.main-thread', stop reason = __llvm_verbose_trap: Function is not implemented
    frame #0: 0x0000000100003fa4 a.out`main [inlined] __llvm_verbose_trap: Function is not implemented at verbose_trap.cpp:0 [opt]
  * frame #1: 0x0000000100003fa4 a.out`main [inlined] Dummy::func(this=<unavailable>) at verbose_trap.cpp:3:5 [opt]
    frame #2: 0x0000000100003fa4 a.out`main at verbose_trap.cpp:8:13 [opt]
    frame #3: 0x0000000189d518b4 dyld`start + 1988
```

(cherry picked from commit 3f3d7aeffe1d156efd4f54cdce6f6c6ca933da75)

### CODE REVIEW

File: lldb/source/Target/VerboseTrapFrameRecognizer.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+lldb::StackFrameSP VerboseTrapRecognizedStackFrame::GetMostRelevantFrame() {
+  return m_most_relevant_frame;
+}
+
+namespace lldb_private {
+
+void RegisterVerboseTrapFrameRecognizer(Process &process) {
+  RegularExpressionSP module_regex_sp = nullptr;
+  RegularExpressionSP symbol_regex_sp(
+      new RegularExpression("(__llvm_verbose_trap)"));
```
Comment (adrian-prantl) [inline lldb/source/Target/VerboseTrapFrameRecognizer.cpp]: should this be `^__llvm_verbose_trap` ?

File: lldb/test/Shell/Recognizer/verbose_trap.test [hunk 2]
```diff
@@ -0,0 +1,9 @@
+# RUN: %clang_host -g -O0 %S/Inputs/verbose_trap.cpp -o %t.out
+# RUN: %lldb -b -s %s %t.out | FileCheck %s
+run
+# CHECK: thread #{{.*}}stop reason = __llvm_verbose_trap: Function is not implemented
```
Comment (adrian-prantl) [inline lldb/test/Shell/Recognizer/verbose_trap.test]: We might consider stripping the `__llvm_verbose_trap` from the user-visible output (and replace it with a more high-level text.

Comment (Michael137) [reply lldb/test/Shell/Recognizer/verbose_trap.test]: Done

File: lldb/source/Target/VerboseTrapFrameRecognizer.cpp (hunk 1 above)
Comment (Michael137) [reply lldb/source/Target/VerboseTrapFrameRecognizer.cpp]: Done

**Comment by Michael137** [comment]
@swift-ci test

**Comment by Michael137** [comment]
@swift-ci test

**Comment by Michael137** [comment]
@swift-ci test

**Comment by Michael137** [comment]
@ahatanaka I cherry-picked your changes in https://github.com/llvm/llvm-project/pull/79230. We can revert and re-cherry-pick once you land your changes on llvm.org

**Comment by Michael137** [comment]
@swift-ci test Windows

**Comment by Michael137** [comment]
@swift-ci test

**Comment by Michael137** [comment]
@swift-ci test

**Comment by Michael137** [comment]
@swift-ci test Windows

**Pull request merged**